control-plane/pkg/reconciler/broker/namespaced_broker.go

```diff
@@ -112,9 +117,56 @@ func (r *NamespacedReconciler) FinalizeKind(ctx context.Context, broker *eventin
 
 	r.IPsLister.Unregister(types.NamespacedName{Namespace: broker.Namespace, Name: broker.Name})
 
+	err := r.cleanUpClusterScopedResources(ctx, broker)
+	if err != nil {
+		// we don't let this error propagate, as we don't want to requeue this forever
+		logging.FromContext(ctx).Errorw("Error while cleaning up cluster scoped resources for KafkaNamespaced brokers in the namespace", zap.Error(err))
+		return result
+	}
+
 	return result
 }
 
+// cleanUpClusterScopedResources deletes the cluster scoped resources created for the data plane in the namespace, if
+// this is the last namespaced broker in the namespace
+func (r *NamespacedReconciler) cleanUpClusterScopedResources(ctx context.Context, broker *eventing.Broker) error {
```

Comment:
The problem with this approach is that it can delete resources for brokers that are created while we're deleting resources.

Reply:
To clarify the sequence of events leading to the race condition:

- broker y deleted in ns x
- list brokers in ns x (as done in cleanUpClusterScopedResources)
- broker z created in ns x
- broker z started reconciliation and created resources
- broker y deletion causes deletion of resources created at the previous phase

However, I see this as an improvement to what we have

Reply:
We could have a mutually exclusive way of created/deleting resources per namespace to solve this race condition

Reply:
Very good catch!

You mean a map of mutexes per namespace? Lock the namespace mutex in during `ReconcileKind` and `FinalizeKind`?

Reply:
I think we are still affected with mutexes per namespace if we ever (or users) enable "bucket based leader election" which might lead to having multiple controller replicas reconciling the same "namespace data plane" but I think that is already better than what we have